// @serenity-js/core/lib/screenplay/interactions/compare_notes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var abilities_1 = require("../abilities");
var CompareNotes = (function () {
    function CompareNotes(actual, expect, topic) {
        var _this = this;
        this.actual = actual;
        this.expect = expect;
        this.topic = topic;
        this.toString = function () { return "{0} compares notes on " + _this.subject(); };
        this.subject = function () { return "" + _this.actual === "" + _this.topic
            ? "" + _this.actual
            : _this.actual + " (noted as \"" + _this.topic + ")"; };
    }
    CompareNotes.toSeeIf = function (question, expectation, topic) {
        return new CompareNotes(question, expectation, topic.toString());
    };
    CompareNotes.prototype.performAs = function (actor) {
        var _this = this;
        return abilities_1.TakeNotes.
            as(actor).
            read(this.topic).
            then(function (expected) { return _this.expect(expected)(actor.toSee(_this.actual)); });
    };
    return CompareNotes;
}());
exports.CompareNotes = CompareNotes;
