// serenity-js/lib/serenity-protractor/stage/photographer-timing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TimingBehaviour = (function () {
    function TimingBehaviour(before, after) {
        this.before = before;
        this.after = after;
    }
    TimingBehaviour.prototype.takeABeforePhoto = function (event, photographer) {
        this.before.takeAPhoto(event.value, event.timestamp, photographer);
    };
    TimingBehaviour.prototype.takeAnAfterPhoto = function (event, photographer) {
        this.after.takeAPhoto(event.value.subject, event.timestamp, photographer);
    };
    return TimingBehaviour;
}());
exports.TimingBehaviour = TimingBehaviour;
var TakeAPhoto = (function () {
    function TakeAPhoto() {
    }
    TakeAPhoto.prototype.takeAPhoto = function (activity, timestamp, photographer) {
        photographer.photograph(activity, timestamp);
    };
    return TakeAPhoto;
}());
exports.TakeAPhoto = TakeAPhoto;
var NoPhoto = (function () {
    function NoPhoto() {
    }
    NoPhoto.prototype.takeAPhoto = function (activity, timestamp, photographer) {
        // no-op
    };
    return NoPhoto;
}());
exports.NoPhoto = NoPhoto;
