// @serenity-js/core/lib/domain/events.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var moment = require("moment");
var DomainEvent = (function () {
    function DomainEvent(value, timestamp) {
        this.value = value;
        this.timestamp = timestamp;
        this.timestamp = timestamp || moment().valueOf();
        this.type = this.constructor.name;
    }
    DomainEvent.prototype.toString = function () {
        return this.timestamp + " | " + this.constructor.name + ": " + this.value.toString();
    };
    return DomainEvent;
}());
exports.DomainEvent = DomainEvent;
var SceneStarts = (function (_super) {
    __extends(SceneStarts, _super);
    function SceneStarts() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SceneStarts;
}(DomainEvent));
exports.SceneStarts = SceneStarts;
var ActivityStarts = (function (_super) {
    __extends(ActivityStarts, _super);
    function ActivityStarts() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ActivityStarts;
}(DomainEvent));
exports.ActivityStarts = ActivityStarts;
var ActivityFinished = (function (_super) {
    __extends(ActivityFinished, _super);
    function ActivityFinished() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ActivityFinished;
}(DomainEvent));
exports.ActivityFinished = ActivityFinished;
var SceneTagged = (function (_super) {
    __extends(SceneTagged, _super);
    function SceneTagged() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SceneTagged;
}(DomainEvent));
exports.SceneTagged = SceneTagged;
var SceneFinished = (function (_super) {
    __extends(SceneFinished, _super);
    function SceneFinished() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SceneFinished;
}(DomainEvent));
exports.SceneFinished = SceneFinished;
var PhotoAttempted = (function (_super) {
    __extends(PhotoAttempted, _super);
    function PhotoAttempted() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return PhotoAttempted;
}(DomainEvent));
exports.PhotoAttempted = PhotoAttempted;
