// @serenity-js/core/lib/screenplay/activities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Task = (function () {
    function Task() {
    }
    return Task;
}());
Task.where = function (description) {
    var activities = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        activities[_i - 1] = arguments[_i];
    }
    return new AnonymousTask(description, activities);
};
exports.Task = Task;
var Interaction = (function () {
    function Interaction() {
    }
    return Interaction;
}());
Interaction.where = function (description, interaction) {
    return new AnonymousInteraction(description, interaction);
};
exports.Interaction = Interaction;
var AnonymousTask = (function () {
    function AnonymousTask(description, activities) {
        var _this = this;
        this.description = description;
        this.activities = activities;
        this.performAs = function (actor) { return actor.attemptsTo.apply(actor, _this.activities); };
        this.toString = function () { return _this.description; };
    }
    return AnonymousTask;
}());
var AnonymousInteraction = (function () {
    function AnonymousInteraction(description, interaction) {
        var _this = this;
        this.description = description;
        this.interaction = interaction;
        this.performAs = function (actor) { return _this.interaction(actor); };
        this.toString = function () { return _this.description; };
    }
    return AnonymousInteraction;
}());
