// serenity-js/lib/serenity-protractor/screenplay/interactions/enter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browse_the_web_1 = require("../abilities/browse_the_web");
var hit_1 = require("./hit");
var Enter = (function () {
    function Enter(value) {
        this.value = value;
        this.interactions = [];
    }
    Enter.theValue = function (value) {
        return new Enter(value);
    };
    Enter.prototype.into = function (field) {
        this.target = field;
        this.interactions.push(new EnterValue(this.value, field));
        return this;
    };
    Enter.prototype.thenHit = function (key) {
        this.interactions.push(hit_1.Hit.the(key).into(this.target));
        return this;
    };
    Enter.prototype.performAs = function (actor) {
        return actor.attemptsTo.apply(actor, this.interactions);
    };
    return Enter;
}());
exports.Enter = Enter;
var EnterValue = (function () {
    function EnterValue(value, target) {
        var _this = this;
        this.value = value;
        this.target = target;
        this.toString = function () { return "#actor enters " + _this.value + " into " + _this.target; };
    }
    EnterValue.prototype.performAs = function (actor) {
        return browse_the_web_1.BrowseTheWeb.as(actor).locate(this.target).sendKeys(this.value);
    };
    return EnterValue;
}());
