// serenity-js/lib/serenity-protractor/screenplay/interactions/resize_browser_window.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browse_the_web_1 = require("../abilities/browse_the_web");
var ResizeBrowserWindow = (function () {
    function ResizeBrowserWindow() {
    }
    return ResizeBrowserWindow;
}());
ResizeBrowserWindow.toMaximum = function () { return new MaximiseBrowserWindow(); };
ResizeBrowserWindow.to = function (width, height) { return new SetBrowserWindowSize(width, height); };
exports.ResizeBrowserWindow = ResizeBrowserWindow;
var MaximiseBrowserWindow = (function () {
    function MaximiseBrowserWindow() {
        this.toString = function () { return "#actor maximises the browser window"; };
    }
    MaximiseBrowserWindow.prototype.performAs = function (actor) {
        return browse_the_web_1.BrowseTheWeb.as(actor).manage().window().maximize();
    };
    return MaximiseBrowserWindow;
}());
var SetBrowserWindowSize = (function () {
    function SetBrowserWindowSize(width, height) {
        var _this = this;
        this.width = width;
        this.height = height;
        this.toString = function () { return "#actor sets the size of the browser window to " + _this.width + " x " + _this.height; };
    }
    SetBrowserWindowSize.prototype.performAs = function (actor) {
        return browse_the_web_1.BrowseTheWeb.as(actor).manage().window().setSize(this.width, this.height);
    };
    return SetBrowserWindowSize;
}());
