// serenity-js/lib/serenity-protractor/screenplay/interactions/select.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var recording_1 = require("@serenity-js/core/lib/recording");
var protractor_1 = require("protractor");
var browse_the_web_1 = require("../abilities/browse_the_web");
var Select = (function () {
    function Select() {
    }
    Select.theValue = function (value) {
        return { from: function (target) { return new SelectOption(value, target); } };
    };
    Select.values = function () {
        var values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            values[_i] = arguments[_i];
        }
        return { from: function (target) { return new SelectOptions(values, target); } };
    };
    return Select;
}());
exports.Select = Select;
var SelectOption = (function () {
    function SelectOption(value, target) {
        var _this = this;
        this.value = value;
        this.target = target;
        this.toString = function () { return "#actor selects \"" + _this.value + "\" from " + _this.target; };
    }
    SelectOption.prototype.performAs = function (actor) {
        return browse_the_web_1.BrowseTheWeb.as(actor)
            .locate(this.target)
            .element(protractor_1.by.cssContainingText('option', this.value))
            .click();
    };
    return SelectOption;
}());
__decorate([
    recording_1.step('{0} selects "#value" from #target'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], SelectOption.prototype, "performAs", null);
var SelectOptions = (function () {
    function SelectOptions(values, target) {
        this.values = values;
        this.target = target;
    }
    SelectOptions.prototype.performAs = function (actor) {
        var _this = this;
        var hasRequiredText = function (option) { return option.getText().then(function (value) { return !!~_this.values.indexOf(value); }); }, isAlreadySelected = function (option) { return option.isSelected(); }, ensureOnlyOneApplies = function (list) { return list.filter(function (_) { return _ === true; }).length === 1; }, select = function (option) { return option.click(); };
        var optionsToClick = function (option) { return protractor_1.protractor.promise.all([
            hasRequiredText(option),
            isAlreadySelected(option),
        ])
            .then(ensureOnlyOneApplies); };
        return browse_the_web_1.BrowseTheWeb.as(actor).locate(this.target)
            .all(protractor_1.by.css('option'))
            .filter(optionsToClick)
            .each(select);
    };
    return SelectOptions;
}());
__decorate([
    recording_1.step('{0} selects "#values" from #target'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], SelectOptions.prototype, "performAs", null);
