// @serenity-js/core/lib/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var Config = (function () {
    function Config(config) {
        this.config = config;
        this.mergeValuesAndConcatenateLists = function (objValue, srcValue) {
            if (_.isArray(objValue)) {
                return objValue.concat(srcValue);
            }
        };
        this.mergeValuesButOverrideLists = function (objValue, srcValue) {
            if (_.isArray(objValue)) {
                return srcValue;
            }
        };
    }
    Object.defineProperty(Config.prototype, "get", {
        get: function () {
            return _.cloneDeep(this.config);
        },
        enumerable: true,
        configurable: true
    });
    Config.prototype.mergedWith = function (otherConfig) {
        return new Config(_.mergeWith({}, this.config, otherConfig, this.mergeValuesAndConcatenateLists));
    };
    Config.prototype.withFallback = function (defaults) {
        return new Config(_.mergeWith({}, defaults, this.config, this.mergeValuesButOverrideLists));
    };
    return Config;
}());
exports.Config = Config;
