// @serenity-js/core/lib/io/file_system.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var gracefulFs = require("graceful-fs");
var mkdirp = require("mkdirp");
var path = require("path");
var FileSystem = (function () {
    function FileSystem(root, fs) {
        if (fs === void 0) { fs = gracefulFs; }
        this.root = root;
        this.fs = fs;
    }
    FileSystem.prototype.store = function (relativePathToFile, data) {
        var _this = this;
        return this.ensureSpecified(relativePathToFile)
            .then(function (relativePath) { return _this.prepareDirectory(relativePath); })
            .then(function (absolutePath) { return _this.write(absolutePath, data); });
    };
    FileSystem.prototype.ensureSpecified = function (relativePathToFile) {
        return !!(relativePathToFile && relativePathToFile.length)
            ? Promise.resolve(relativePathToFile)
            : Promise.reject(new Error('Please specify where the file should be saved'));
    };
    FileSystem.prototype.prepareDirectory = function (relativePathToFile) {
        var _this = this;
        var absolutePath = path.resolve(this.root, relativePathToFile), parent = path.dirname(absolutePath);
        return new Promise(function (resolve, reject) {
            mkdirp(parent, { fs: _this.fs }, function (error) {
                return error
                    ? reject(error)
                    : resolve(absolutePath);
            });
        });
    };
    FileSystem.prototype.write = function (path, data) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fs.writeFile(path, data, function (error) {
                return error
                    ? reject(error)
                    : resolve(path);
            });
        });
    };
    return FileSystem;
}());
exports.FileSystem = FileSystem;
